File: freecell/data_provider.py
import configparser
import os.path

from freecell import Statistics, INPUT_FILE


class DataProvider:
    """Reads raw data from the configuration file that contains the history"""

    def __init__(self):
        """Creates a new DataProvider"""
        statistic: str = DataProvider.get_statistic_string()
        wins, total, best, worst = [int(x)
                                    for x in statistic.split(";")
                                    if x]
        self._statistics = Statistics(wins, total, best, worst)

    @property
    def statistics(self) -> Statistics:
        """Returns the Statistics object"""
        return self._statistics

    @staticmethod
    def get_statistic_string() -> str:
        """Loads the configuration from the default input file.
        This method can be replaced by a mock object in unit tests"""

        # Read the freecell configuration
        input_file = INPUT_FILE
        if not os.path.exists(input_file):
            errmsg = f"'{input_file}' configuration file not found"
            raise RuntimeError(errmsg)

        # Load a configuration parser object from it
        config = configparser.ConfigParser()
        config.read(input_file)

        # Read the statistics string from the freecell.scm section
        section = config["freecell.scm"]
        if not section:
            errmsg = f"configuration does not contain a freecell section"
            raise RuntimeError(errmsg)

        stat_string = config["freecell.scm"]["Statistic"]
        if not stat_string:
            errmsg = f"Section 'freecell.scm' does not contain 'Statistic'"
            raise RuntimeError(errmsg)

        return stat_string

File: freecell/data_provider.go
package freecell

import (
	"os"
	"path/filepath"

	ini "github.com/ochinchina/go-ini"
)

const (
	CONFIG_FILE_NAME = ".config/gnome-games/aisleriot"
)


// Loads the configuration from the default input file. This method can
// be replaced by a mock object in unit tests
//
// A configuration file provider is a function returning either:
//  1. A file name (such as the actual configuration file)
//  2. A string containing the contents of a configuration file
// The go-ini package will accept either one.
//
// The default configuration file provider gets the name of
// the actual AisleRiot configuration file in the home directory.
func GetStatisticString(provider ...func() string) (string, error) {

	// Get a function that will provide the contents of the config file
	var configFileProvider func() string

	// If no alternate provider was found, use the one relative to
	// the actual home directory.  Otherwise, use the provider specified
	// on the command line
	switch {
	case len(provider) == 0:
		configFileProvider = func() string {
			home, _ := os.UserHomeDir()
			configFile := filepath.Join(home, CONFIG_FILE_NAME)
			return configFile
		}
	default:
		configFileProvider = provider[0]
	}

	// Use that function to load the configuration .ini file
	configFile := configFileProvider()
	config := ini.Load(configFile)

	// Get the freecell section and the statistic string from it
	var statisticString string
	section, err := config.GetSection("freecell.scm")
	if err == nil {
		statisticString, err = section.GetValue("Statistic")
	}
	return statisticString, err
}
